### DISCUSSION TITLE
How to configure several authentication systems ?

### DISCUSSION THREAD

**Discussion by ghost** [adonisjs/core#3302] [Help] [answered]
Hi, i have the following scenario: 
1. First i have the admin system which requires the admin to login and do stuff that admins can only do like creating products etc.
2. Second i have the customers system where the customers should login and do things that customers only can do like viewing their orders, buy some stuff etc.

I am fine in defining 1 authentication system but i am stuck in creating 2. I have admin model and customer model, along with their migrations and tables, also their controllers. 

The part i want help with is the stuff in the `contracts/auth.ts` and `config/auth.ts`. How can i configure both files to match my scenario ? 

Note that they both use authentication tokens cuz i am using an API and i have a separate front-end vue app, also i am using a sql databse to store the tokens and i think there should be 2 tables for storing tokens, one for customers tokens and one for admins tokens.

**Comment by ammezie** [comment, marked as answer]
One way of doing this is, you can create new gaurds (simply duplicating the `api` guard) like below:

```
  admin: {
      driver: 'oat',

      tokenProvider: {
        driver: 'database',
        table: 'admin_api_tokens',
        foreignKey: 'user_id',
      },

      provider: {
        identifierKey: 'id',
        uids: ['email'],
        model: () => import('App/Models/Admin'),
      },
  },
customer: {
      driver: 'oat',

      tokenProvider: {
        driver: 'database',
        table: 'customer_api_tokens',
        foreignKey: 'user_id',
      },

      provider: {
        identifierKey: 'id',
        uids: ['email'],
        model: () => import('App/Models/Customer'),
      },
  }
```
Then you can do:
```
auth.use('admin').login()
auth.use('customer').login()
```

The way I personally would go about this is introduce the concept of roles. So there will be 2 roles: admin and customer. That way, you will have a single "user" model/table. It can be as simple as just adding a `role` column to the users table. Then depending on users role, you can display whatever is appropriate.

**Reply by ghost** [reply]
Actually the approach you mentioned of having one users table and then define roles in a much better approach.
